source/hic_hal/nxp/lpc55xx/uart.c

```diff
@@ -189,15 +195,17 @@ int32_t uart_write_data(uint8_t *data, uint16_t size)
         return 0;
     }
 
-    uint32_t cnt = 0;
-    if (circ_buf_count_used(&write_buffer) > 0) {
+    // Disable interrupts to prevent the uart_handler from modifying the
+    // circular buffer at the same time.
+    NVIC_DisableIRQ(USART_IRQ);
+    uint32_t cnt = circ_buf_write(&write_buffer, data, size);
+    if (cb_buf.tx_size == 0 && circ_buf_count_used(&write_buffer) > 0) {
```

Comment:
given that the circular buffer has calls to disable/enable interrupts I *think* that I might be able to remove the NVIC_DisableIRQ and NVIC_EnableIRQ calls as this is written (I'm leaning towards I think I can remove them but still need to convince myself). Note that the `circ_buf_count_used(&write_buffer) > 0` part is needed even if we know that we just added >0 bytes to the circular buffer in that case. What do you think?

Reply:
Well, ideally yes you could remove the NVIC IRQ disable/enable calls due to the circ_buf interrupt masking. But it looks like there is a race condition in `circ_buf_write()` in how it doesn't wrap the whole call in interrupt masking and instead relies on `circ_buf_push()` protecting each byte insert, with only one check for MIN() and no for loop exit on overflow. So for now it's probably best to keep the IRQ disabling as you have it.

Reply:
(Added #841 for the general `circ_buf_write()` issue.)